Add interaction constraint shortcuts to `HistGradientBoosting*`

betatim opened the pull request:
Allow users to specify common interaction constraints as a string. Users can specify `"pairwise"` or `"no interactions"` as a shortcut to having to explicitly provide the interactions.


#### Reference Issues/PRs
Closes #24845

betatim committed:
Add interaction constraint shortcuts

Allow users to specify common interaction constraints as a string.

betatim committed:
Update doc strings with interaction cst shortcut

betatim committed:
Add whats new entry

betatim commented:
This is ready for review.

amueller commented:
I think it would be awesome to have an example with "no interactions" and discuss how that leads to an additive model, which is therefore super-interpretable, and say that it's very closely related to the explainable boosting machine (EBM) of Caruana (but slightly different).
For pairwise interactions, I think it would be nice to maybe have a separate example that explains what it means. Honestly I don't have a good intuition, and I think it would be good to explain that it's not an additive model on pairwise interactions (if I remember correctly).

amueller commented:
Hm I'm thinking now that with all pairwise constraints, each leaf is an indicator with two features, and a tree is a weighted sum of these indicators, so it is actually possible to rewrite the function in this way, just in a less obvious way than for EBMs?

If that's the case that would be another interesting unit test maybe?

amueller commented:
This could be a separate PR, but I would suggest breaking up https://scikit-learn.org/dev/auto_examples/inspection/plot_partial_dependence.html#sphx-glr-auto-examples-inspection-plot-partial-dependence-py into two examples, and showing that with the interaction constraints the predictions are  actually the sum of the partial dependencies, both for the univariate case and the pairwise interaction case.

ogrisel commented on sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py:
```diff
@@ -261,29 +262,43 @@ def _check_interaction_cst(self, n_features):
-                f"got: {self.interaction_cst!r}."
-            )
+        if isinstance(self.interaction_cst, str):
+            if self.interaction_cst == "no interactions":
```
I don't think we have any string hyper-parameter values in scikit-learn with a white space in them. Intuitively I would use replace the whitespace with an underscore but I did not do a consistency review of current hyper-parameter values in the code base.

It might be possible to conduct such a review programmatically thanks to `sklearn.utils.all_estimator` and the `_parameter_constraints` attributes of each estimator class.

betatim replied:
Ah, I have no strong feelings on this. I used a whitespace because I thought it is "by humans, for humans" - it just looks prettier to me like this.

```python
In [15]: for estimator_name, klass in sklearn.utils.all_estimators():
    ...:     if not hasattr(klass, "_parameter_constraints"):
    ...:         continue
    ...:     for name, constraints in klass._parameter_constraints.items():
    ...:         for constraint in constraints:
    ...:             if isinstance(constraint, sklearn.utils._param_validation.StrOptions):
    ...:                 print(estimator_name, name, constraint.options)
```
says you are right, all the options so far use `_`.

betatim replied:
Switched to using an underscore. And `StrOptions` in the `_parameter_constraints`.

ogrisel reviewed (commented):
Besides the concern with the white space. LGTM.

I would also like at least a quick update of an existing example (or new example) to show the "no interaction" which might be quite common.

A full-blown example for EBM-style glassbox modeling might come in a follow-up PR.

amueller commented:
>  Subtracting the mean y before fitting might "fix" the discrepancy.

The example removes the mean of y on the whole dataset. I can try removing just the training set mean and not using early stopping, but even without early stopping, the difference doesn't seem directly related to the mean of ``y``.

I was gonna try and empirically confirm the glass box assertion was also true for the pairwise model but got sidetracked with meetings :-/

lorentzenchr commented:
@amueller Side note: Github has $\LaTeX$ ($) support now.
Main note: If I‘m not mistaken, specifying pairwise interaction constraints gives a model that is pairwise additive in link space: $m(x) = h(\sum_i f_i(x_i) + \sum_{i\lt j} f_{i,j}(x_i,x_j))$

betatim committed:
Switch from whitespace to underscore for consistency

Renamed parameter value to use an underscore and use `StrOptions` to
describe valid parameter values.

betatim commented:
> > I think it would be awesome to have an example with "no interactions" and discuss how that leads to an additive model, which is therefore super-interpretable, and say that it's very closely related to the explainable boosting machine (EBM) of Caruana (but slightly different).
> 
> I think this can be done as part of the current PR, right? That would be nice to link to such an example from the release highlights.

I'll look at adding an example.

I'd prefer to punt all the other ideas about changes to examples or new examples to a separate PR (or PRs). "hashtag one idea per PR"

lorentzenchr commented:
@betatim I agree. I also prefer a new PR for an additional example to keep this PR small.

betatim commented:
Any recommendations for a good didactic dataset/setup for an example? Otherwise I'll do a bit of research to find something.

lorentzenchr reviewed (commented):
Looks very good. 2 comments:
- Should we add a test that models are identical for string options and their manual pendent?
- In the original PR for interaction constraints, concerns were raised that "pairwise" expands to a list with $n(n-1)/2$ elements. For 10000 features this is not great. One could use iterators/generators inside the code to improve memory footprint, but I would not do so as long as the need does not arise.

betatim commented on sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py:
```diff
@@ -261,29 +262,43 @@ def _check_interaction_cst(self, n_features):
-            )
+            elif self.interaction_cst == "pairwise":
+                interaction_cst = itertools.combinations(range(n_features), 2)
+            else:
+                raise ValueError(
+                    f"'{self.interaction_cst}' is not a valid interaction constraint. "
+                    "Use 'no_interactions', 'pairwise' or specify them explicitly."
+                )
```
Did you mean to ask "I think this is captured by `StrOptions`, so we don't need this?"

betatim replied:
If yes, then I think the answer is that I only discovered `StrOptions` this morning. Still getting used to the brave new world with such input validation support (compared to having to write it yourself back in the day).

betatim commented:
> * In the original PR for interaction constraints, concerns were raised that "pairwise" expands to a list ..

I saw that, not sure there is much we can do about it. A few lines later in `_check_interaction_cst` the code compuates `constraints = [set(group) for group in interaction_cst]` which will consume any iterator/generator. So the memory will get used there. As there is nothing to stop users from passing in an equivalent pairwise interaction constraint, I wouldn't stop them from using the shortcut.

Rough estimate: 1000 features, 10000 samples leads to ~80000000 bytes (`np.ones((1000, 10_000), dtype=np.float64).nbytes`) for the input data and ~4167352 bytes (`sys.getsizeof([set(g) for g in itertools.combinations(range(1000), 2)])`) for the pairwise constraints.

In the case of 10000 features it takes a considerable amount of time to even run `sys.getsizeof([set(g) for g in itertools.combinations(range(1000), 2)])`, long enough to write this sentence and still have to wait. It uses about 400MB of memory. I think to fix this you'd have to replace the list of sets with some other data structure to represent all the constraints, or restructure the code so that you can pass the string `"pairwise"` through to the low(est) levels of the tree building code.

So I'd wait and see how many people have datasets with 10000 features and for whom using ~500MB of memory to represent the constraints is a problem. My guess is that most people with that many features use a large amount of memory just for their data, so another 500MB here or there is a rounding error in the total budget.

betatim committed:
Update test to use new parameter value

betatim committed:
Remove explicit parameter validation

betatim committed:
Update doc string

betatim commented on sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py:
```diff
@@ -1246,6 +1261,9 @@ class HistGradientBoostingRegressor(RegressorMixin, BaseHistGradientBoosting):
         they are treated as if they were specified as an additional set.
 
+        Alternatively, "pairwise" or "no_interactions" are shorthands for
+        allowing only pairwise/no interactions.
```
What do you think of "The strings "pairwise" and "no_interactions" are shorthands for allowing only pairwise or no interactions, respectively.". It allowed me to remove the "alternatively" which I didn't like.

lorentzenchr replied:
That's good.

lorentzenchr commented:
As said earlier, I would wait if memory problems with `interaction_cst="pairwise` arise at all. (BTW, it only scales with `n_features`, not with `n_samples`).
If this need arises, a way to go may be to pass the string `"pairwise"` instead of a list of sets to all helper functions for interaction constraints.

lorentzenchr commented on doc/whats_new/v1.2.rst:
```diff
@@ -289,6 +289,11 @@ Changelog
 
+- |Enhancement| Interaction constraints for :class:`~sklearn.ensemble.HistGradientBoostingClassifier`
+  and :class:`~sklearn.ensemble.HistGradientBoostingRegressor` can now be specified
+  as strings for two common cases: no interactions and pairwise interactions.
```
Agreed. How about
```
... two common cases: `"no_interactions"` and `"pairwise"` interactions.
```

betatim replied:
Works for me.

betatim committed:
Update PDP example to use "no_interactions" shortcut

betatim commented:
The reason I pick a (random) value for `n_samples` was to get an estimate of how much memory users are using to hold their data. Gives a sense of scale to how much extra memory the list uses. If your dataset is 16GB, then 400MB is nothing. If the dataset only consumed 40MB then an additional 400MB is a big deal.

betatim committed:
Update whats new

lorentzenchr commented:
@betatim Could you fix the linting errors (activating pre-commit hooks might help with black).

betatim committed:
Fix formatting

betatim commented:
Formatting fixed, before we merge this we should look at (and decide on a way to resolve) #24899.

amueller commented:
> @amueller If I‘m not mistaken, specifying pairwise interaction constraints gives a model m that is pairwise additive in link space: m(x)=h(∑ifi(xi)+∑i<jfi,j(xi,xj)). The functions f, however, are not identical to the trees. It may be the case that different branches (always including the root) of a tree belong to different f.

That was my understanding as well :) so the main interpretability aspect is still there, but it's "just different" from the EBMs. I think it's worth saying both of these in the new example in the new PR :)

betatim committed:
Fix doc string

betatim committed:
Small fixes

betatim commented:
Updated now that #24899 is merged. I think all the comments about examples are meant for a new PR, this means I think this is ready.

ogrisel reviewed (approved):
LGTM!

ogrisel commented:
I will try to fix the broken test locally.

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/betatim/24849

jeremiedbb committed:
right place in what's new

jeremiedbb committed:
remove test for error because it's handled by the common test

jeremiedbb committed:
nit

jeremiedbb committed:
follow docstring conventions

jeremiedbb commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -1,3 +1,4 @@
+import itertools
```
```suggestion
```

jeremiedbb committed:
Update sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py

jeremiedbb commented:
I directly pushed the required fixes for the CI and some nitpicks. We can remove the test checking for the error message when invalid string for the parameter. This is checked by a common test.

jeremiedbb reviewed (approved):
LGTM

jeremiedbb commented:
Thanks @betatim

lorentzenchr commented:
Very cool to have this PR in 1.2. Thanks @betatim and everyone involved.
